amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -110,10 +121,84 @@ public void restartFrom(RecordsRetrieved recordsRetrieved) {
+            }
         }
     }
 
+    private boolean evictAckedEventAndScheduleNextEvent(RecordsRetrievedAck recordsRetrievedAck,
+            CompletableFuture<RecordFlow> triggeringFlowFuture) {
+        boolean isNextEventScheduled = false;
+        // Remove the head of the queue on receiving the ack.
+        // Note : This does not block wait to retrieve an element.
+        RecordsRetrievedContext recordsRetrievedContext = recordsDeliveryQueue.poll();
+        // Check if the ack corresponds to the head of the delivery queue.
+        if (recordsRetrievedContext != null && recordsRetrievedContext.getRecordsRetrieved().batchUniqueIdentifier()
+                .equals(recordsRetrievedAck.batchUniqueIdentifier())) {
+            // Update current sequence number for the successfully delivered event.
+            currentSequenceNumber = recordsRetrievedAck.deliveredSequenceNumber();
+            // Update the triggering flow for post scheduling upstream request.
+            triggeringFlowFuture.complete(recordsRetrievedContext.getRecordFlow());
+            // Try scheduling the next event in the queue, if available.
+            if (recordsDeliveryQueue.peek() != null) {
+                subscriber.onNext(recordsDeliveryQueue.peek().getRecordsRetrieved());
+                isNextEventScheduled = true;
+            }
+        } else {
+            // TODO : toString implementation for recordsRetrievedAck
+            log.error("{}: KCL BUG: Found mismatched payload {} in the delivery queue for the ack {}  ", shardId,
+                    recordsRetrievedContext.getRecordsRetrieved(), recordsRetrievedAck);
+            throw new IllegalStateException("KCL BUG: Record delivery ack mismatch");
+        }
+        return isNextEventScheduled;
+    }
+
+    private boolean bufferCurrentEventAndScheduleIfRequired(RecordsRetrieved recordsRetrieved, RecordFlow triggeringFlow) {
+        boolean isCurrentEventScheduled = false;
+        final RecordsRetrievedContext recordsRetrievedContext = new RecordsRetrievedContext(recordsRetrieved, triggeringFlow);
+        try {
+            // Try enqueueing the RecordsRetrieved batch to the queue, which would throw exception on failure.
+            // Note: This does not block wait to enqueue.
+            recordsDeliveryQueue.add(recordsRetrievedContext);
+            // If the current batch is the only element in the queue, then try scheduling the event delivery.
+            if (recordsDeliveryQueue.size() == 1) {
+                subscriber.onNext(recordsRetrieved);
+                isCurrentEventScheduled = true;
+            }
+        } catch (IllegalStateException e) {
+            log.warn("{}: Unable to enqueue the payload due to capacity restrictions in delivery queue with remaining capacity {} ",
+                    shardId, recordsDeliveryQueue.remainingCapacity());
+            throw e;
+        } catch (Throwable t) {
+            recordsDeliveryQueue.remove(recordsRetrievedContext);
```

Comment:
In the context of this method, it failed to schedule the event that was already queued. Hence we remove it from queue. But when we throw back the throwable, it will be caught by the errorOccured and the queue will be cleared.

Reply:
We can decide on whether we need to log an error.